### PR TITLE
Add raycast-ia-writer extension

### PR DESCRIPTION
**Pull request by xvvhang**
## Description

There are two commands in this extension to let users could create or search notes quickly inside iA Writer library.

With Quick Note command, users could jot down anything and specify the filename of their notes or use time stamp as filenames.
With Quick Search command, users could open iA Writer's Quick Search window with query instantly.

The extension is made with iA Writer's URL schemes.

## Screencast

quick-note

quick-search

## Checklist

- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [x] I ran `npm run build` and [tested this distribution build in Raycast](https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration)
- [x] I checked that files in the `assets` folder are used by the extension itself
- [x] I checked that assets used by the `README` are placed outside of the `metadata` folder

### CODE REVIEW

**Review by pernielsentikaer — CHANGES_REQUESTED** [review]
Hi 👋 

Thanks for your contribution 💪 

I have now tested your extension, and I have some feedback ready for you:

- Maybe you can check if Ai Writer is installed? You can see how extensions like `Eagle` does it 🙂

- Why is current datetime populated in the form, is that something most people wants?

- Could you add at least one [screenshot](https://developers.raycast.com/basics/prepare-an-extension-for-store#screenshots) of the extension

I'm looking forward to testing this extension again 🔥 

Request a new review when you are ready. Feel free to contact me here or at [Slack](https://www.raycast.com/community) if you have any questions.

**Comment by xvvhang** [comment]
> Hi 👋
> 
> Thanks for your contribution 💪
> 
> I have now tested your extension, and I have some feedback ready for you:
> 
> * Maybe you can check if Ai Writer is installed? You can see how extensions like `Eagle` does it 🙂
> * Why is current datetime populated in the form, is that something most people wants?
> * Could you add at least one [screenshot](https://developers.raycast.com/basics/prepare-an-extension-for-store#screenshots) of the extension
> 
> I'm looking forward to testing this extension again 🔥
> 
> Request a new review when you are ready. Feel free to contact me here or at [Slack](https://www.raycast.com/community) if you have any questions.

Hi

Thanks for your review. I’ve got your feedback and made some changes:

1. Added installation check like what `Eagle` extension did.
2. Added screenshots to `/metadata` folder.

And I know your confusion about the datetime populated in Quick Note form. It’s a common feature in some note apps like Obsidian([Unique note creator](https://help.obsidian.md/Plugins/Unique+note+creator)), which makes it easy to create quick notes in a [Zettelkasten](https://en.wikipedia.org/wiki/Zettelkasten) way. There are options to disable the feature or specify datetime format. So users can use this extension in their ways.

Looking forward to more feedbacks.
Thank you!

File: extensions/raycast-ia-writer/src/quick-note.tsx [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    <Form
+      actions={
+        <ActionPanel>
+          <Action.SubmitForm onSubmit={onSubmit} />
+          <Action title="Reset Form" onAction={resetForm}></Action>
+        </ActionPanel>
+      }
+    >
+      <Form.TextField id="name" title="Filename" value={name} onChange={(name) => setName(name)} />
+      <Form.TextArea id="text" title="Content" value={text} onChange={(text) => setText(text)} />
```
Comment (pernielsentikaer) [inline extensions/raycast-ia-writer/src/quick-note.tsx:46]:
```suggestion
      <Form.TextArea
        id="text"
        title="Content"
        value={text}
        autoFocus={defaultText() !== ""}
        onChange={(text) => setText(text)}
      />
```
Maybe set autofocus when date is set?

**Review by pernielsentikaer — CHANGES_REQUESTED** [review]
Hi 👋

I tested the new improvements and have some feedback 🙂

I still see "Note Saved" if I enters the wrong credentials, do you have an idea why?

**Comment by xvvhang** [comment]
> Hi 👋
> 
> I tested the new improvements and have some feedback 🙂
> 
> I still see "Noe Saved" if I enters the wrong credentials, do you have an idea why?

Hi  👋

The Quick Note command calls mac's `open` command with iA Writer's URL Scheme in a node child process. The `open` command runs successfully even there's something wrong with URL's query string. So I couldn't know if the note is created or not. 🥲

I checked how `Bear` and `Obsidian` deal with the case. `Bear` just opens the URL without any success toasts. And `Obsidian` creates a new file or overwrites the file if it's existed.

I prefer the `Bear` way and may remove the `Note Saved` toast. If the note's not saved correctly. iA Writer just pop up. What do you think of it ? Should I remove the toast? 🤔

Thanks for your feedback!

**Comment by pernielsentikaer** [comment]
Let's do that, this was we don't confuse the end user 🙂

**Comment by mil3na** [comment]
Hi!

I'm taking a look at your extension while Per is on vacay :D 
By first running it, without IA Writer installed, I can see an error on the bottom of the screen. However, I can still see the inputs and ⌘Enter "saves the note" successfully. I don't believe this would be the best approach to this. The first screen should ask the user all the things needed for the extension to work properly. 
Do you mind changing that?

Also, the quick search is not working for me. It does nothing other than show a HUD.

**Comment by xvvhang** [comment]
> Hi!
> 
> I'm taking a look at your extension while Per is on vacay :D By first running it, without IA Writer installed, I can see an error on the bottom of the screen. However, I can still see the inputs and ⌘Enter "saves the note" successfully. I don't believe this would be the best approach to this. The first screen should ask the user all the things needed for the extension to work properly. Do you mind changing that?
> 
> Also, the quick search is not working for me. It does nothing other than show a HUD.

Hi,

I just removed the wrong toast when iA writer is not installed and updated the icon.
As Per asked, I checked how `Eagle` extension works when Eagle app not installed. It just shows a toast asking users to install Eagle app. And I think this extension does it in the same way. Could you please check how other extensions work when apps not installed like `Obsidian` and `Bear`?

Thanks for your review!

**Comment by mil3na** [comment]
Hey! 
I understand that Per suggested it, but I don't think it is a good UI. It is confusing for the user. You could do the same check you are doing now, but show a different "screen" with instructions for the user, in case they don't have the app installed.

**Comment by xvvhang** [comment]
> Hey! I understand that Per suggested it, but I don't think it is a good UI. It is confusing for the user. You could do the same check you are doing now, but show a different "screen" with instructions for the user, in case they don't have the app installed.

Hi,

I understand your concern. But I wonder why there isn't a 'different' screen in `Bear` or `Obsidian` extension when I don't have Bear or Obsidian installed? Both of them are published to the store and got thousands of downloads. I don't really understand.

If it's required to add such a 'different' screen, should I use an `Alert` or push to a new `Detail` page? Could you please give me some best practices examples to learn from?

Thanks for your review.

**Comment by pernielsentikaer** [comment]
We could do something like this @xvvhang 🙂

raycast-ia-writer 2024-04-03 at 16 28 29

You're right there is other extensions in the store that doesn't handle it but might - most people installing this extension should have AI Writer so it's not the worst issue - could you check my additions?

**Comment by xvvhang** [comment]
> We could do something like this @xvvhang 🙂
> 
> raycast-ia-writer 2024-04-03 at 16 28 29
> 
> You're right there is other extensions in the store that doesn't handle it but might - most people installing this extension should have AI Writer so it's not the worst issue - could you check my additions?

Sure! That make senses and I'm glad to make some changes. I will commit some changes later. Thank you Per!🫡

**Comment by pernielsentikaer** [comment]
I already made above in the extension, can you approve it? 🙂

**Review by xvvhang — COMMENTED** [review]
LGTM!

**Comment by xvvhang** [comment]
Thanks for your help! 🙂@pernielsentikaer

**Review by pernielsentikaer — APPROVED** [review]
Hi 👋 

Looks good to me, approved 🔥

**Pull request merged**